includes/collection/class-inbox.php

```diff
@@ -140,4 +141,93 @@ public static function get( $guid, $user_id ) {
 
 		return \get_post( $post_id );
 	}
+
+	/**
+	 * Get an inbox item by its GUID.
+	 *
+	 * @param string $guid The GUID of the inbox item.
+	 *
+	 * @return \WP_Post|\WP_Error The inbox item or WP_Error.
+	 */
+	public static function get_by_guid( $guid ) {
+		global $wpdb;
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+		$post_id = $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT ID FROM $wpdb->posts WHERE guid=%s AND post_type=%s",
+				\esc_url( $guid ),
+				self::POST_TYPE
+			)
+		);
+
+		if ( ! $post_id ) {
+			return new \WP_Error(
+				'activitypub_inbox_item_not_found',
+				\__( 'Inbox item not found', 'activitypub' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		return \get_post( $post_id );
+	}
+
+	/**
+	 * Undo a received activity.
+	 *
+	 * @param string $id The ID of the inbox item to be removed.
+	 *
+	 * @return bool|\WP_Error True on success, WP_Error on failure.
+	 */
+	public static function undo( $id ) {
+		$type = '';
+		$post = self::get_by_guid( $id );
+
+		if ( \is_wp_error( $post ) ) {
+			// If inbox entry not found, return the error.
+			return $post;
+		}
+
+		$type = \get_post_meta( $post->ID, '_activitypub_activity_type', true );
+
+		switch ( $type ) {
+			case 'Follow':
+				$actor   = \get_post_meta( $post->ID, '_activitypub_activity_remote_actor', true );
+				$user_id = $post->post_author;
+				$post    = Remote_Actors::get_by_uri( $actor );
+
+				if ( \is_wp_error( $post ) ) {
+					return $post;
+				}
+
+				return Followers::remove( $post, $user_id );
+			case 'Like':
+			case 'Create':
+			case 'Announce':
+				if ( ACTIVITYPUB_DISABLE_INCOMING_INTERACTIONS ) {
+					return new \WP_Error(
+						'activitypub_inbox_undo_interactions_disabled',
+						\__( 'Undo is not possible because incoming interactions are disabled.', 'activitypub' ),
+						array( 'status' => 403 )
+					);
+				}
+
+				$result = Comment::object_id_to_comment( esc_url_raw( $post->guid ) );
+
+				if ( empty( $result ) ) {
+					return new \WP_Error(
+						'activitypub_inbox_undo_comment_not_found',
+						\__( 'Undo is not possible because the comment was not found.', 'activitypub' ),
+						array( 'status' => 404 )
+					);
+				}
+
+				return \wp_delete_comment( $result, true );
+			default:
+				return new \WP_Error(
+					'activitypub_inbox_undo_unsupported',
+					\__( 'Undo is only supported for Follow, Like, Create, and Announce activities.', 'activitypub' ),
```

Comment:
I don't know if we'll ever expand Undo support, but instead of hard-coding the list of supported types that might change, we could do the inverse and declare the lack of support for the current type. Something like  `sprintf( 'Undo is not supported for %s activities.', $type )`

Reply:
makes sense!